consensus/consensus.go

```diff
@@ -82,8 +80,77 @@ func (fc *FastHotStuffConsensus) Init(
 	return nil
 }
 
-func (fc *FastHotStuffConsensus) UpdateChainTip( /* TODO */ ) {
-	// TODO
+// UpdateChainTip is called when the caller has connected a new block at the tip of the blockchain,
+// for an arbitrary block height and view. Given the new block has been connected, this function
+// resets the internal timers and state of the Fast HotStuff consensus that determine the next action.
+func (fc *FastHotStuffConsensus) UpdateChainTip(chainTip Block, validators []Validator) error {
+	// Grab the consensus instance's lock
+	fc.lock.Lock()
+	defer fc.lock.Unlock()
+
+	// Ensure the consensus instance is running
+	if fc.status != consensusStatusRunning {
```

Comment:
We do. I've updated to use the `fc.IsRunning()` function now

Reply:
yep - this is fine then!

Reply:
So, we had a similar situation in the mempool code. We arrived at not holding a lock in `IsRunning`; however, this runs into possible race conditions with read/write to the status.

Feels like this `IsRunning` rabbit hole is prevalent in the concurrent code, so it would be nice to find a clear-cut solution that we can apply to other components. We have possible solutions:

1. Ditch `IsRunning` and hold a general lock inside the function.
2. Add an `isRunningNoLock` variant and hold a general lock inside the function.
3. (NEW) Add a `statusMutex` only held during read/write access to the status.

Option 1. seems the easiest to implement and reason about, and it's also what Sofonias done here. But option 3 gives us the advantage of having public-facing functions like `IsRunning`. So I think option 3 is interesting to consider. What do you guys think? Which option do you prefer? It would be nice to have a standardized solution to this.

Reply:
The IsRunning and IsInitialized function are no longer needed. I've removed them.